contracts/Utils.sol

```diff
@@ -23,9 +38,13 @@ contract Utils {
 
     mapping(IERC20 => uint256) internal decimals;
 
-    function getUpdateDecimals(IERC20 token) internal returns (uint256) {
-        if (token == ETH_TOKEN_ADDRESS) return ETH_DECIMALS; // save storage access
-        uint256 tokenDecimals = decimals[token];
+    function updateDecimals(IERC20 token) internal returns (uint256) {
```

Comment:
isn't this
getUpdateDecimals.

since we want it to also be the getter of non view functions.

Reply:
Yes. But I still feel it's confusing if we have the "get" prefix. So I thought, the main reason they call updateDecimals() is to set the decimals, but just so happens to return the token decimals as well. Since there's somewhat a duplication, I thought of this instead:

```solidity
function updateDecimals(IERC20 token) internal returns (uint256 tokenDecimals) {
        tokenDecimals = getDecimals(token);
        decimals[token] = tokenDecimals;
}

function tokenSwitch(IERC20 token) internal pure returns (uint256) {
        if (token == ETH_TOKEN_ADDRESS) {
            return ETH_DECIMALS;
        } else if (token == USDT_TOKEN_ADDRESS) {
            return 6;
        } else if (token == DAI_TOKEN_ADDRESS) {
            return 18;
        } else if (token == USDC_TOKEN_ADDRESS) {
            return 6;
        } else if (token == WBTC_TOKEN_ADDRESS) {
            return 8;
        } else if (token == KNC_TOKEN_ADDRESS) {
            return 18;
        } else {
            return 0;
        }
 }

function getDecimals(IERC20 token) internal view returns (uint256 tokenDecimals) {
        // return token decimals if handled by token switch
        tokenDecimals = tokenSwitch(token);
        if (tokenDecimals > 0) return tokenDecimals;

        // handle case where token decimals is not a declared constant
        tokenDecimals = decimals[token];
        // moreover, very possible that old tokens have decimals 0
        // these tokens will just have higher gas fees.
        return (tokenDecimals > 0) ? tokenDecimals : token.decimals();
}
```

We can possibly change the name  to `getSetDecimals()`.

Reply:
The caveat of the above is that we are storing to decimals map the constant token decimals, even though it will never be accessed to returned. But that is only if someone calls `updateDecimals()` using any of the constant token addresses.

Reply:
1. 
Thing is I do want us to use getUpdateDecimals when possible not only for updating. also for getting on non view functions.

2. seems to me updateDecimals should only update if the token isn't in constant values.